### PR TITLE
REF: de-duplicate _wrap_joined_index in MultiIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
Follow-up to #36282

We'll also be able to clean up MultiIndex special-casing in _reindex_non_unique.  I expect others but cant say for sure yet.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3876,7 +3876,10 @@ def _join_monotonic(self, other, how="left", return_indexers=False):
             return join_index
 
     def _wrap_joined_index(self, joined, other):
-        name = get_op_result_name(self, other)
+        if isinstance(self, ABCMultiIndex):
+            name = self.names if self.names == other.names else None
+        else:
+            name = get_op_result_name(self, other)
```
Comment (jreback) [inline pandas/core/indexes/base.py:3882]: hmm why can't we do this in get_op_result_name?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3882]: i considered that, decided against it because it meant additional overhead in a _lot_ of places

Comment (jreback) [reply pandas/core/indexes/base.py:3882]: kk

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -152,6 +153,25 @@ def _codes_to_ints(self, codes):
         return np.bitwise_or.reduce(codes, axis=1)
 
 
+def names_compat(meth):
```
Comment (jreback) [inline pandas/core/indexes/multi.py:156]: type if you can

**Review by jreback — APPROVED** [review]
assume you are going to use names_compat in more places

**Pull request merged**